Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts

```diff
@@ -0,0 +1,190 @@
+        throw new Error("No workbooks in workbook package.");
+
+    // Grab index.workbook or the first workbook file if there isn't an index.
+    const workbookFile = workbookFiles.find(wf => wf.name === "index.workbook") || workbookFiles[0];
+    const workbookFileContents = await workbookFile.async("text")
+
+    // TODO: Pass down _all_ of the files so that we can implement #r and #load.
+    return await loadWorkbookFromString(workbookSession, workbookPackage.name, workbookFileContents)
+}
+
+export interface WorkbookManifest {
+    packages: List<PackageDescription>
+    platforms: List<WorkbookTarget>
+    title?: string
+    id?: string
+    uti: string
+    rest: {}
+}
+
+export class Workbook {
+    private readonly _rawMetadata: any
+    private readonly _markdownContent: string
+    private readonly _manifest: WorkbookManifest
+    private readonly _fileName?: string
+
+    get rawMetadata(): any {
+        return this._rawMetadata
+    }
+
+    get markdownContent(): string {
+        return this._markdownContent
+    }
+
+    get manifest(): WorkbookManifest {
+        return this._manifest
+    }
+
+    get fileName(): string | undefined {
+        return this._fileName
+    }
+
+    constructor(workbookSession: WorkbookSession, fileName: string, markdownContent: string, metadata: any) {
+        this._fileName = fileName;
+        this._markdownContent = markdownContent.trim()
+        this._rawMetadata = metadata
+        this._manifest = this.parseManifest(this._rawMetadata, workbookSession)
+    }
+
+    getManifestToSave(): any {
+        const saveableManifest: any = Object.assign({
+            id: this.manifest.id,
+            uti: this.manifest.uti,
+            title: this.manifest.title,
+        }, this.manifest.rest)
+        saveableManifest.packages = this.manifest.packages.map(pd => {
+            return !pd ? undefined : { id: pd.packageId, version: pd.identityVersion }
+        }).filter((_: any) => _).toArray()
+        saveableManifest.platforms = this.manifest.platforms.map(wt => {
+            return !wt ? undefined : getPlatformFromWorkbookTargetId(wt.id)
+        }).filter((_: any) => _).toArray()
+        return saveableManifest
+    }
+
+    parseManifest(rawMetadata: any, workbookSession: WorkbookSession): WorkbookManifest {
+        let checkUti = true;
+        const finalMetadata: any = { uti: defaultUti }
+
+        if (!rawMetadata) {
+            rawMetadata = {}
+            checkUti = false
+        } else {
+            // Make a copy so we can modify it.
+            rawMetadata = Object.assign({}, rawMetadata)
+        }
+
+        const hasMatchingUti = (!rawMetadata.uti || rawMetadata.uti.localeCompare(defaultUti, undefined, { sensitivity: "base" }));
+        if (checkUti && hasMatchingUti)
+            throw new Error("Invalid manifest");
+        else
+            delete rawMetadata.uti
+
+        let id = rawMetadata.id;
+        if (id && isUUID.v4(id))
```

Comment:
I think it's sufficient to simply check that it's just not null/empty. We definitely don't care what version the GUID may be, or even if it's a GUID. The only reason it's a v4 GUID is because that's all that `System.Guid` will produce via `NewGuid`. We don't even use it anywhere<sup>1</sup>. You can eliminate the `is-uuid` dependency.

_<sup>1.</sup> This originated on the first pass of pages where it was going to be used for referencing. If we ever get back to that, it may not even survive._